crates/provider/src/new.rs

```diff
@@ -91,13 +155,223 @@ impl<N, T> RootProviderInner<N, T> {
 #[cfg_attr(not(target_arch = "wasm32"), async_trait::async_trait)]
 #[auto_impl::auto_impl(&, &mut, Rc, Arc, Box)]
 pub trait Provider<N: Network, T: Transport + Clone = BoxTransport>: Send + Sync {
+    /// Returns the root provider.
+    fn root(&self) -> &RootProvider<N, T>;
+
     /// Returns the RPC client used to send requests.
-    fn client(&self) -> ClientRef<'_, T>;
+    ///
+    /// NOTE: this method should not be overridden.
+    #[inline]
+    fn client(&self) -> ClientRef<'_, T> {
+        self.root().client()
+    }
+
+    /// Returns a [`Weak`](std::sync::Weak) RPC client used to send requests.
+    ///
+    /// NOTE: this method should not be overridden.
+    #[inline]
+    fn weak_client(&self) -> WeakClient<T> {
+        self.root().weak_client()
+    }
+
+    /// Watch for the confirmation of a single pending transaction with the given configuration.
+    ///
+    /// Note that this is handled internally rather than calling any specific RPC method, and as
+    /// such should not be overridden.
+    #[inline]
+    async fn watch_pending_transaction(
+        &self,
+        config: PendingTransactionConfig,
+    ) -> TransportResult<PendingTransaction> {
+        self.root().watch_pending_transaction(config).await
+    }
+
+    /// Subscribe to new block headers.
+    ///
+    /// Returns a stream of new block headers.
+    #[cfg(feature = "pubsub")]
+    async fn subscribe_blocks(&self) -> TransportResult<Subscription<Block>> {
+        self.root().pubsub_frontend()?;
+        let id = self.client().prepare("eth_subscribe", ("newHeads",)).await?;
+        self.root().get_subscription(id).await
+    }
+
+    /// Subscribe to new pending transactions.
+    ///
+    /// Returns a stream of new pending transaction hashes.
+    #[cfg(feature = "pubsub")]
+    async fn subscribe_pending_transactions(&self) -> TransportResult<Subscription<B256>> {
+        self.root().pubsub_frontend()?;
+        let id = self.client().prepare("eth_subscribe", ("newPendingTransactions",)).await?;
+        self.root().get_subscription(id).await
```

Comment:
nit this makes 2 calls to pubsub_frontend due to the get_subscription(id) call, but should be fine

Reply:
yes, we don't want to make the request if we know we're not a pubsub